<a href="https://colab.research.google.com/github/jcjimenezb123/MetodosNumericosPython/blob/master/fractal_newton.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

El método de Newton-Raphson es un método abierto para resolver ecuaciones no lineales. Resolver se entiende como encontrar los valores de x donde la función f(x) es cero
$$
f(x)=0
$$
El valor particular de x donde la función es cero se conoce como raíz.

El método de Newton-Raphson se deriva de la serie de Taylor, donde se aproxima una función por medio de la serie y esa serie se iguala a cero, para encontrar la aproximación a la raíz se despeja de la serie truncada de primer orden

$$
0=f(x)=f(x_0)+f'(x_0)(x-x_0)+...
$$

Donde $x_0$ es un valor inicial con la que se hace la primer aproximación y $f'(x)$ es la derivada de $f(x)$.
El programa es sencillo:

In [2]:
def nr(f,df,x0,maxit=100,tol=1e-6):
  k=0 #valor inicial de iteraciones
  x=x0 
  while k<maxit:#mientras las iteraciones sean menor a la maxima
    x = x - f(x)/df(x) #aplica el metodo de Newton
    k = k + 1 #incrementa la iteracion
    if abs(f(x))<tol: #verifica el criterio de convergencia
        return x, #<---------- retorna la raiz y las iteraciones
  return x, #<---------- retorna la raiz y las iteraciones

Donde *maxit* es el número máximo de iteraciones en el ciclo, *k* es el contador de iteraciones, *tol* es el valor máximo que puede tener la función para aceptarse como la raíz de $f(x)$. Ambos valores son arbitrarios y se dan de acuerdo al problema que se quiere resolver.

Por ejemplo, si se quiere saber la raíz de la siguiente función

$$
f(x)=x^2-2
$$

obtenemos su derivada que es

$$
f'(x)=2x
$$

Podemos dar un valor inicial por ejemplo $x_0=1$, un valor máximo de iteraciones $= 100$ y una tolerancia $= 1e-5$. 

Ejecutamos el programa de Newton-Raphson para encontrar la raíz

In [14]:
f  = lambda x: #<---- defina la funcion
df = lambda x: #<---- defina su derivada
x0       = #valor inicial
niter    = #numero maximo de iteraciones
tol      = #tolerancia

raiz,k=nr(f,df,x0,niter,tol) #se llama el metodo y devuelve la raiz y las iteraciones
print(raiz,) #<---- mostrar la raiz encontrada y el numero de iteraciones

1.4142156862745099 3


El resultado es raiz$=1.4142$ en $3$ iteraciones. Si cambiamos el valor inicial por ejemplo $x_0=0.5$ obtenemos

In [4]:
x0       = #<---- cambie el valor inicial

raiz,k=nr(f,df,x0,niter,tol)
print(raiz,k)

1.4142135625249321 5


Observamos que llegamos al mismo resultado de la raíz pero con $5$ iteraciones, ahora probemos con otro valor de $x_0=0.1$

In [5]:
x0       = #<----- cambie el valor inicial

raiz,k=nr(f,df,x0,niter,tol)
print(raiz,k)

1.4142136001158032 7


Sigue obteniendo el resultado pero ahora en $7$ iteraciones, hagamos otra prueba, ahora con $x_0=0.001$

In [6]:
x0       = #<---- cambie el valor inicial

raiz,k=nr(f,df,x0,niter,tol)
print(raiz,k)

1.4142135626178514 14


Observamos que igualmente llega al resultado pero ahora le toma $14$ iteraciones, notamos que si el valor inicial se acerca a 0 entonces le toma más iteraciones llegar a la raíz, note que no es posible que $x_0=0$ porque la derivada es cero y el valor de x se indetermina y no es posible obtener la raíz.

In [13]:
x0       = #<---- cambie el valor inicial

raiz,k=nr(f,df,x0,niter,tol)
print(raiz,k)

1.4142135623974939 54


Podemos pensar que existen valores de $x_0$ donde es más fácil encontrar la raíz y existen otros donde encontrar la raíz es un poco más tardado o puede divergir.

Seamos un poco curiosos y aberigüemos qué comportamiento tiene el método de Newton-Raphson para otros valores iniciales. Si hacemos un proceso donde cambiemos el valor inicial de $x_0$ en el intervalo $[-2, 2]$ y obtenemos el número de iteraciones $k$ que le toma llegar a la raíz en cada caso y grafiquemos esos resultados para observar el comportamiento

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

n=100
rangox=[, ] #<---- asigne los valores del intervalo
xs=np.linspace(rangox[0],rangox[1],n)
ks=np.zeros(n)
xr=np.zeros(n)
for j in range(n):
    x0=xs[j]
    xr[j],ks[j]=nr(f,df,x0,niter,tol)

plt.plot(xs,ks,xs,xr)
plt.title('Newton-Raphson para f(x)=x^2-2')
plt.xlabel('x inicial')
plt.ylabel('iteraciones')
plt.grid()


Observamos dos mínimos cercanos a $-1.4142$ y $1.4142$, esto nos indica que se requieren menos iteraciones en valores cercanos a estos puntos.
Será el mismo comportamiento para otras funciones?

Cambiemos ahora la función a:

$$
f(x)=x^3-1
$$
$$
f'(x)=3x^2
$$

In [ ]:
f  = lambda #<----- defina la funcion
df = lambda #<----- defina su derivada
niter = 50
tol = 1e-5
import numpy as np
import matplotlib.pyplot as plt

n=100
rangox=[-2, 2]
xs=np.linspace(rangox[0],rangox[1],n)
ks=np.zeros(n)
xr=np.zeros(n)
for j in range(n):
    x0=xs[j]
    xr[j],ks[j]=nr(f,df,x0,niter,tol)

plt.plot(xs,ks,xs,xr)
plt.title('Newton-Raphson para f(x)=x^3-1')
plt.xlabel('x inicial')
plt.ylabel('iteraciones')
plt.grid()

A diferencia de la función cuadrática anterior, ésta función cúbica tiene 3 raíces, una es real y las otras dos son complejas. El método de Newton-Raphson es capaz de encontrar raíces complejas, solo que el valor inicial $x_0$ debe ser un valor complejo. Entonces vamos a crear un valor complejo variariando tanto el valor real como el imaginario para crear un *lienzo* con los ejes **real-imaginario**

Para poder graficar tanto el valor real como el imaginario y además el número de iteraciones, tendremos que usar una gráfica en tres dimensiones donde el plano x-y es el valor complejo y el eje z son las iteraciones.

In [ ]:
import numpy as np
import cmath
import matplotlib.pyplot as plt
from mpl_toolkits import mplot3d

n=60
lienzo = np.zeros((n,n))
rangox=[-2, 2]
rangoy=[-2, 2]
xs=np.linspace(rangox[0],rangox[1],n) #valores reales
ys=np.linspace(rangoy[0],rangoy[1],n) #valores imaginarios
[x, y] = np.meshgrid(xs, ys) #combinaciones de cada valor

ks=np.zeros(n)
for i in range(n):
  for j in range(n):
    x0=complex(x[i,j],y[i,j]) #se crea el valor complejo
    [xr,ks]=nr(f,df,x0)
    lienzo[i,j]=  #<--- crea el lienzo para cada nodo coloca el numero de iteraciones

fig = plt.figure()
ax = plt.axes(projection='3d')

ax.plot_surface(x, y, lienzo,cmap='viridis', edgecolor='none')
ax.set_title('Newton-Raphon f(x)=x^3-1')
ax.set_xlabel('Parte real')
ax.set_ylabel('Parte imaginaria')
ax.set_zlabel('Iteraciones')
plt.show()

Observamos zonas de rapida convergencia son las más bajas y zonas de lenta convergencia que son las más altas. Ahora cambiemos la representación gráfica y en lugar que sea una gráfica en tres dimensiones la vamos a crear en 2 dimensiones donde x es la parte real, y es la parte imaginaria y el número de iteraciones es el color del punto

In [ ]:
import numpy as np
import cmath
import matplotlib.pyplot as plt

n=900
lienzo = np.zeros((n,n))
rangox=[-2, 2]
rangoy=[-2, 2]
xs=np.linspace(rangox[0],rangox[1],n) #valores reales
ys=np.linspace(rangoy[0],rangoy[1],n) #valores imaginarios
[x, y] = np.meshgrid(xs, ys) #combinaciones de cada valor

ks=np.zeros(n)
for i in range(n):
  for j in range(n):
    x0=complex(x[i,j],y[i,j]) #se crea el valor complejo
    [xr,ks]=nr(f,df,x0)
    lienzo[i,j]=ks

fig,ax=plt.subplots(1,1)
cp = ax.contourf(x, y, ) #<---- graficar el lienzo

plt.show()

A éste gráfico se le conoce como el Fractal de Newton-Raphson. Probemos ahora con la siguiente función

In [ ]:
import numpy as np
import cmath

n=1000
lienzo = np.zeros((n,n))
rangox=[-2, 2]
rangoy=[-2, 2]
xs=np.linspace(rangox[0],rangox[1],n) #valores reales
ys=np.linspace(rangoy[0],rangoy[1],n) #valores imaginarios
[x, y] = np.meshgrid(xs, ys) #combinaciones de cada valor

from PIL import Image
image = Image.new(mode='L', size=(n, n))

for i in range(n):
  for j in range(n):
    x0=complex(x[i,j],y[i,j]) #se crea el valor complejo
    [xr,ks]=nr(f,df,x0)
    image.putpixel((i, j), *15) #<---- dibujar el lienzo

image.show()

In [ ]:
import numpy as np
import cmath

n=1000
lienzo = np.zeros((n,n))
rangox=[-2, 2]
rangoy=[-2, 2]
xs=np.linspace(rangox[0],rangox[1],n) #valores reales
ys=np.linspace(rangoy[0],rangoy[1],n) #valores imaginarios
[x, y] = np.meshgrid(xs, ys) #combinaciones de cada valor

from PIL import Image
image = Image.new(mode='RGB', size=(n, n))

f = lambda x:x**6+x**3-1
df = lambda x:6*x**5+3*x**2

for i in range(n):
  for j in range(n):
    x0=complex(x[i,j],y[i,j]) #se crea el valor complejo
    [xr,ks]=nr(f,df,x0)
    image.putpixel((i, j), *25555) #<---- colocar el numero de iteraciones

image.show()